// @app/libs/middlewares/jwt-api-auth.js
const url = require('url');
const querystring = require('querystring');
const jwt = require('jsonwebtoken');
const debug = require('debug')('jwtAuth');
module.exports = jwtAuth;

function json(data, code) {
  if(!code) code = 200;
  this.res.writeHead(code, {
    'Content-Type': 'application/json'
  });
  var status = 'success';
  if(code >= 400) status = 'error';

  var mess = null;
  if(data.message) {
    mess = data.message + '';
    delete data.message;
  }
  var obj = {
    status: status,
    data: data
  };
  if(mess != null) obj.message = mess;

  this.res.end(JSON.stringify(obj));
}

async function setInReq(req, res, next) {
  res.json = json.bind({res: res, next: next});
  this.req = req;
  this.res = res;
  this.next = next;
  req.jwtAuth = this;
  var uri = url.parse(req.url);
  var qs = querystring.parse(uri.query);
  var token;
  
  const bearerToken = req.headers.authorization;
  const xAuthToken = req.headers['x-auth-token'] || null;
  const getToken = qs['authToken'] || null;
  const cToken = req.cookies['authToken'] || null;
  //const pathname = url.parse(req.url).pathname;

  if(bearerToken) token = bearerToken.replace('Bearer ', '');
  if(xAuthToken) token = xAuthToken;
  if(getToken) token = getToken;
  if(cToken) token = cToken;
  debug('jwt', token);
  if(token) {
    req.jwt = token;
    try {
      req.jwtData = jwt.verify(token, this.opts.secret);
    } catch (e) {
      debug('error', e.message);
      req.jwtError = e.message;
      if(e.message.indexOf('expired')) {
        req.cookies = {};
      }
    } 
  }
  if(this.opts.privateApi) {
    var conf = this.opts.api || {};
    if(!conf.base) conf.base = 'api';
    if(!conf.pub) conf.pub = [];
    if(this.req.url.indexOf(`/${conf.base}/`) ===  0 && conf.pub.indexOf(this.req.url) < 0) {
      try{
        await this.isRequired();
      } catch (e) {
        debug('isRequired:error', e);
      } 
    }
  }
  next();
    
}
function makeToken(payload) {
  var token = jwt.sign(payload, this.opts.secret, {expiresIn: this.opts.expiration});
  var info = jwt.decode(token);
  info.authToken = token;
  return info;
}

async function isRequired() {
  if(!this.req.jwt) {
    this.res.json({message: 'auth token not found.'}, 401);
  }
  if(!this.req.jwtData) {
    var message = 'auth data not found.';
    if(this.req.jwtError) {
      message = `auth token issue: ${this.req.jwtError}`;
    }
    this.res.json({message: message}, 401);
  }
  return;
}
/* throw Error inside cb to respond as error */
async function allowIf(cb) {
  try {
    await this.isRequired();
    await cb(this.req.jwtData);
  } catch (e) {
    this.res.json({message: e.message}, 401);
  }
}

function jwtAuth(opts) {
  if (!(this instanceof jwtAuth)) return new jwtAuth(opts);
  this.opts = Object.assign({expiration: '1hr', privateApi: false}, opts || {});
  this.makeToken = makeToken;
  this.isRequired = isRequired;
  this.allowIf = allowIf;
  debug('opts', this.opts);
  return setInReq.bind(this);
}
